Conventions settings

fberlakovich opened the pull request:
#### Summary of additions and changes

I have finally started working on the possibility to configure certain conventions (e.g. which commands/environments should have labels and which prefixes should be used). Before continuing with the GUI, however, I would like to hear your thoughts.

What I envision:
* A table that  allows configuring which commands/environments should receive labels and how these labels should look like. That is, which elements are governed by the missing label/label convention inspections.
* Possibly settings for other conventions that are currently hard coded
* Like the code style settings, the conventions should be configurable on the project level or on the IDE level, possibly even with multiple schemes/templates available at the IDE level.

What this pull request currently provides:
* A basic infrastructure to store settings on the project/IDE level
* A draft of a table to configure label convntions

![image](https://user-images.githubusercontent.com/7955528/118624189-375bba80-b7c9-11eb-8af2-967607e4f7e5.png)

Known issues:
* Currently only the maximum section size settings are actually persisted
* None of the settings are actually used yet by any inspection
* Currently no new IDE level schemes can be created

In a first version the table would not allow the addition or deletion of rows because we cannot currently  detect whether labels should be added as a command or as an optional parameter (see the related Gitter discussion). @PHPirates @HannahSchellekens  @stenwessel  Do you have any thoughts on the table? Do you even think a table is the right UI element for that purpose?
Also the maximum section size spinner just serves as an example. Whether it is actually a good idea to make this convention configurable is open for discussion. Do you have any thoughts on which settings should be configurable?

I also want to mention that a lot of code and implementation ideas were taken from https://github.com/FWDekker/intellij-randomness. I will definitely mention the authort in the comments once the code matures.

HannahSchellekens commented:
Thanks for the effort!

> Do you have any thoughts on the table? Do you even think a table is the right UI element for that purpose?
Also the maximum section size spinner just serves as an example. Whether it is actually a good idea to make this convention configurable is open for discussion. Do you have any thoughts on which settings should be configurable?

I do think that this is a good idea. As much as we love our conventions that we gathered over time, we must acknowledge that everyone has a different workflow and that it can be valuable to share these conventions with your whole team. The inspection suppression/magic comments was a first step in this.

To see which functionality could be configured, take a look at the [`nl.hannahsten.texifyidea.util.magic`](https://github.com/Hannah-Sten/TeXiFy-IDEA/tree/master/src/nl/hannahsten/texifyidea/util/magic) package. These are all magic values that could be configured (not all of them are suited). And there are TONS of them. The magic constants were there to centralize all magic values to make migrating to configurable settings easier while still keeping development momentum. It also wouldn't surprise me if some magic values are hidden in code as well...

The control elements you've chosen for the amount of characters seem right. However, we have to properly think about how we're going to present all settings. Having too many settings in one place is daunting for newer users. [You don't want this](https://youtu.be/dKx1wnXClcI?t=747). A few questions to ask ourselves:

- How do we keep some sense of overview?
- How do we make it easy for new users to understand the settings?
- What settings make sense to be configurable?
- What will be the hierarchy?
- What settings are more important, what settings can be hidden away?
- How can you make the settings page intuitive?

A similar issue to this is #1713 (run configuration ui).

> on the project/IDE level

To me, project level seems like the correct one. Especially when collaborating, you want everyone on the same page. However, I'd also consider adding global defaults.

Also, JetBrains has a [number spinner](https://github.com/Hannah-Sten/TeXiFy-IDEA/pull/1917/files#diff-01a6c0262f2e932ffe18845a42b2aa0dfe0a4af256480de9bc7ad9183a036471) implemented already. They provide a [JBIntSpinner](https://github.com/JetBrains/intellij-community/blob/master/platform/platform-api/src/com/intellij/ui/JBIntSpinner.java). I used it in the Insert dummy text dialog:

![image](https://user-images.githubusercontent.com/17410729/118634073-6dea0300-b7d2-11eb-8688-e33e3a5820cf.png)

You can use [internal mode](https://plugins.jetbrains.com/docs/intellij/enabling-internal.html) and select the UI Inspector `Tools` > `Internal Actions` > `UI` > `UI Inspector` to look for suitable elements in the UI.

fberlakovich commented:
> I do think that this is a good idea. As much as we love our conventions that we gathered over time, we must acknowledge that everyone has a different workflow and that it can be valuable to share these conventions with your whole team. The inspection suppression/magic comments was a first step in this.

I agree. My particular use case was that I had to use different label defaults for certain projects because I had to obey the project guidelines. I think the project level / IDE level scheme system of IntelliJ is a good match for such a use case.


> To see which functionality could be configured, take a look at the [`nl.hannahsten.texifyidea.util.magic`](https://github.com/Hannah-Sten/TeXiFy-IDEA/tree/master/src/nl/hannahsten/texifyidea/util/magic) package. These are all magic values that could be configured (not all of them are suited). And there are TONS of them. The magic constants were there to centralize all magic values to make migrating to configurable settings easier while still keeping development momentum. It also wouldn't surprise me if some magic values are hidden in code as well...

I think that one constraint should be that "conventions" should not allow for an incorrect configuration. That is, there should be no case where configuring a setting in a certain way leads to e.g. syntactically or semantically incorrect code fixes, inspections etc. When adhering to this constraint, many magic values would require automatic detection of additional command/environment properties (e.g. is a newly added labeled command labeled with a \label command or an optional parameter).

> The control elements you've chosen for the amount of characters seem right. However, we have to properly think about how we're going to present all settings. Having too many settings in one place is daunting for newer users. [You don't want this](https://youtu.be/dKx1wnXClcI?t=747). A few questions to ask ourselves:

I have updated the PR so the table settings are actually persisted and the initial values reflect the label prefixes in the magic constants. Also, the enabled label conventions reflect the "Minimum sectioning level which should trigger the missing label inspection". As such, the table could replace the mentioned setting and the `labeledPrefixes` magic constants. Currently the settings for environments and commands are combined in one table. We could split them into two tables, however. What do you think?

At a quick glance the only other magic constant that could be made configurable easily is `sectionSeparatorColors`. However, I don't really see a use case for it for now. Also the maximum section size spinner was more of a demo than an actual use case. Should we maybe limit this PR to 
1) the infrastructure to create project vs. IDE wide convention settings and 
2) the configuration of enabled label inspections and prefixes?

With these two pieces in place, adding more settings should be considerably less effort. 

> Also, JetBrains has a [number spinner](https://github.com/Hannah-Sten/TeXiFy-IDEA/pull/1917/files#diff-01a6c0262f2e932ffe18845a42b2aa0dfe0a4af256480de9bc7ad9183a036471) implemented already. They provide a [JBIntSpinner](https://github.com/JetBrains/intellij-community/blob/master/platform/platform-api/src/com/intellij/ui/JBIntSpinner.java). I used it in the [Insert dummy text dialog](https://github.com/Hannah-Sten/TeXiFy-IDEA/blob/a78fba4fface22f315c040dc50c308110f0f50dc/src/nl/hannahsten/texifyidea/action/wizard/ipsum/InsertDummyTextDialogWrapper.kt#L33):

Fixed, thank you.

fberlakovich commented:
@HannahSchellekens do I need to mention you or the other reviewers in my comments for you to receive notifications? (If you already were notified, please don't feel rushed. I am just not sure how these draft PRs work).

PHPirates commented:
Is this relevant? https://blog.jetbrains.com/idea/2021/05/intellij-platform-ui-kit/ I don't understand what it is though, we already have https://jetbrains.design/intellij/

fberlakovich commented:
> Is this relevant? https://blog.jetbrains.com/idea/2021/05/intellij-platform-ui-kit/ I don't understand what it is though, we already have https://jetbrains.design/intellij/

I also cannot figure out what it is meant to be. It seems like a marketing pitch for the UI guidelines to me.

fberlakovich commented:
What are your thoughts on the questions regarding the state of this PR in my comment (specifically, how do you like the table prototype and what about the scope of this PR)? I am slightly hesistant to continue working in a certain direction without your confirmation that you are happy with the ideas.

PHPirates commented:
@fberlakovich As a general comment, I think a nice feature of TeXiFy is that it actually _has_ UI elements at all for configuring settings, that are at least sort of intuitively understandable, not all LaTeX editors have that.
So it will be specific for each configurable setting what a suitable UI element is, for example in this case a table could well be it.

I did not check the code right now as there's a lot more to do here, but of course ideally we should make it such that it's easy to add more UI elements for magic, however this is still in an early stage, so right now we maybe should wait with generalising until we have more information on how we can make things easier and more generic from a programming point of view.

I have to check whether the TexifyConventionsConfigurable makes sense to me, I will come back to that later. You could help me by writing class documentation about the what and why.

fberlakovich commented:
> I have to check whether the TexifyConventionsConfigurable makes sense to me, I will come back to that later. You could help me by writing class documentation about the what and why.

Sure, I will add documentation asap.

fberlakovich commented:
@PHPirates sorry for the delay, I added documentation to the classes. Unfortunately, the relation between the classes is still quite confusing, but I gave my best to explain it in the comments. I took some of the ideas regarding the settings classes from https://github.com/FWDekker/intellij-randomness. The settings view is currently limited to a small number of settings to demonstrate the principle, but with the general infrastructure in place can easily be expanded to host other project/global settings. Feel free to contact me if you have any questions.

PHPirates commented on src/nl/hannahsten/texifyidea/settings/conventions/TexifyConventionsConfigurable.kt:
```diff
@@ -0,0 +1,233 @@
+ *
+ * The parameters need default values for XML serialization.
+ */
+data class LabelConvention(
```
You can put this in a separate file

PHPirates commented on src/nl/hannahsten/texifyidea/settings/conventions/TexifyConventionsConfigurable.kt:
```diff
@@ -0,0 +1,233 @@
+                }
+            }
+
+        val enabledColumnInfo = object : TableModelEditor.EditableColumnInfo<LabelConvention, Boolean>("") {
```
I think this should have a name as well, it's not clear what it is enabling (the inspection which check prefixes? whether it generates a prefix automatically or not?)
Maybe we should add a help text to clarify

fberlakovich replied:
I added a column name (still looks a bit ugly, because the checkbox is centered) as well as a tooltip. The idea behind the tooltip was to control the `LatexMissingLabelInspection`. Such a control would also supersede the "minimum section level which triggers the missing label inspection" because with the checkbox a user could decide to have labels on a specific section level but not another one. If you prefer the current setting though we can also repurpose the checkbox to control the `LatexLabelConventionInspection` inspection instead or even have two checkboxes.

PHPirates replied:
Ah, well, that's why a column name was necessary I guess :D No I think the current purpose is fine.

Yes, of course the spacing should be fixed, and I think the column should be on the right instead of the left (so that the column with elements for which you change settings, is on the left)

PHPirates commented on src/nl/hannahsten/texifyidea/settings/conventions/TexifyConventionsGlobalSettingsManager.kt:
```diff
@@ -0,0 +1,37 @@
+ * settings remember which scheme is currently selected. Instances of this class must be serializable, since they are
+ * persisted by a [TexifyConventionsGlobalSettingsManager].
+ */
+data class TexifyConventionsGlobalState(
```
So the idea is, that calls to Magic classes are replaced by calls to the scheme state (via the settings manager, the project one if you have it and otherwise global) to get the appropriate scheme?

fberlakovich replied:
Yes, or more precisely: you should always use the `TexifyConventionsSettingsManager` since otherwise you won't know which scheme is active because you can't determine it from the `TexifyConventionsGlobalState` alone. The `TexifyConventionsSettingsManager` returns a `TexifyConventionsSettings` object which determines the active scheme (`currentScheme`) automatically. To do so, it needs both, a  `TexifyConventionsProjectState` and a `TexifyConventionsGlobalState` because for each project whether a project scheme is used or not is stored in the `TexifyConventionsProjectState`.

fberlakovich replied:
Once the open questions are resolved I can port the label inspections to use the convention settings. That should make everything clearer.

PHPirates replied:
Makes sense to me, I don't think there are many cases where you would need access to conventions settings without knowing which project it is for.

PHPirates replied:
>  I can port the label inspections to use the convention settings

Yes, please do :)

PHPirates commented on src/nl/hannahsten/texifyidea/settings/conventions/TexifyConventionsSettingsManager.kt:
```diff
@@ -0,0 +1,90 @@
+
+    companion object {
+
+        fun getInstance(project: Project) =
```
This facade claims to be useful for automatically figuring out whether global or project settings need to be used, but you need a project to get an instance?

fberlakovich replied:
Yes, because paradoxically the project instance is needed to decide whether a project scheme is active (see my explanation above).

PHPirates commented on src/nl/hannahsten/texifyidea/settings/conventions/TexifyConventionsSettingsManager.kt:
```diff
@@ -0,0 +1,90 @@
+ * the currently active scheme modifies either the project state or the global state, depending on which scheme is
+ * selected.
+ */
+data class TexifyConventionsSettings(
```
There's many classes related to schemes and settings, many of which are not in different files, which can make it harder to find them.
To have a better overview, can you think of a structure to make it clearer which classes are intended to be called outside this package (mainly to get currently applicable conventions), and which ones are 'package internal' abstractions?

fberlakovich replied:
I made everything package internal `internal`, reducing the public API to `TexifyConventionsSettings` and `TexifyConventionsSettingsManager` basically.

PHPirates replied:
Yes, that should be clear enough

PHPirates commented on src/nl/hannahsten/texifyidea/settings/conventions/TexifyConventionsScheme.kt:
```diff
@@ -0,0 +1,70 @@
+    /**
+     * The maximum section size before the corresponding inspection issues a warning.
+     */
+    var maxSectionSize: Int = 4000,
```
So this class specifies the default values?

fberlakovich replied:
Exactly, I added a hint in the documentation.

PHPirates commented on src/nl/hannahsten/texifyidea/settings/conventions/TexifyConventionsScheme.kt:
```diff
@@ -0,0 +1,70 @@
+ * Instances must be serializable since they are persisted as part of [TexifyConventionsGlobalState] or
+ * [TexifyConventionsProjectState].
+ */
+data class TexifyConventionsScheme(
```
It's great to have schemes, I think that will work much better than the separate project settings that we used to have.

It looks like the implementation is very targeted towards conventions settings. Why is that, isn't this more generally applicable to the other TeXiFy settings (which are currently global settings) as well?

fberlakovich replied:
> It's great to have schemes, I think that will work much better than the separate project settings that we used to have.

I am glad you like it. 

> It looks like the implementation is very targeted towards conventions settings. Why is that, isn't this more generally applicable to the other TeXiFy settings (which are currently global settings) as well?

The only reason it is currently targeted towards conventions is that they are the first use case I could think of. The principle is definitely applicable to other TeXiFy settings as well. One would have to think carefully in which contexts a project/global scheme or more generally a scheme A/scheme B/scheme C separation makes sense. That is, which settings are global, which ones vary from project to project (e.g. IntelliJ's inspection schemes) and which ones should have completely independent schemes (e.g. IntelliJ's colour schemes).

Most of the conventions settings code can be generalized with generics, leaving the conventions as an instantiation of a more general infrastructure. If you think that is something that could be useful in the future I can extract general interfaces and classes from the existing code.

PHPirates replied:
No, don't generalise more than necessary. It's just good to have in mind where it could be applicable to use this, but for now let's don't make it more complicated than it already is.

PHPirates commented on src/nl/hannahsten/texifyidea/settings/conventions/TexifyConventionsConfigurable.kt:
```diff
@@ -0,0 +1,233 @@
+        mainPanel.add(schemesPanel, BorderLayout.NORTH)
+        mainPanel.add(centerPanel, BorderLayout.CENTER)
+        loadSettings(settingsManager.getSettings())
+        return mainPanel
```
Now, we should carefully consider which settings should be on this page. I think the "minimum section level which triggers the missing label inspection" should be here, because it recommends certain LaTeX conventions, and other settings can remain where they are.

fberlakovich replied:
The reason I didn't include the "minimum section level which triggers the missing label inspection" setting was because of my intended meaning of the checkbox (see my explanation above).

The `sectionSeparatorColors` would be an possible candidate for the conventions though. Alternatively we could inlcude them in the "Editor -> Color Scheme" like other languages do.

PHPirates replied:
Yes, that makes my comment redundant (except that we shouldn't forget to remove that setting there). 
I would think colors should be in the color scheme, yes.

PHPirates reviewed (commented):
Yes, let's make this the UI in front of the magic we have, I think that is good. It would be great to work with schemes, that was definitely missing.

You can further integrate it now :)
The UI might need some polishing, but I think it's good as it is, no need to make it more complicated than necessary.

PHPirates commented:
Awesome! Let me know and I'll be ready to review it

PHPirates commented on src/nl/hannahsten/texifyidea/settings/conventions/TexifyConventionsConfigurable.kt:
```diff
@@ -0,0 +1,216 @@
+    private lateinit var maxSectionSize: JBIntSpinner
+    private lateinit var labelConventionsTable: TableModelEditor<LabelConvention>
+
+    private fun createConventionsTable(): TableModelEditor<LabelConvention> {
```
The table UI still has the plus/minus/copy buttons, but they aren't functional so can be removed?
It would of course be better if they were functional, but for me that's a nice to have and not necessary right now.

fberlakovich replied:
I couldn't find a way to remove it and also not the copy button. The copy button in particular is added in `createComponent`, but I can't simply override it because to create the component you need the `toolbarDecorator` which is a private member of the `TableModelEditor` unfortunately.

As for editing in general: yes, that would be the non plus ultra, but I outlined in my first comment why that would be a major effort. To add new conventions we would need to know how the added commands/environments work syntactically (e.g. whether the label is a parameter or a command). Since we don't have a way yet to figure it out automatically, the only alternative would be for the user the enter it, which bears the risk that the entered information might be incorrect, leading to incorrect modifications.

PHPirates replied:
I removed the buttons, please check if I didn't break anything else

fberlakovich replied:
> I removed the buttons, please check if I didn't break anything else

Thanks, looks good to me.

PHPirates commented on src/nl/hannahsten/texifyidea/settings/conventions/TexifyConventionsConfigurable.kt:
```diff
@@ -0,0 +1,216 @@
+            override fun getItemClass(): Class<out LabelConvention> = LabelConvention::class.java
+
+            override fun clone(item: LabelConvention, forInPlaceEditing: Boolean): LabelConvention {
+                TODO("Not yet implemented")
```
to do

fberlakovich replied:
I changed it to a simple exception. It is not implemented on purpose though (see above)

PHPirates commented on src/nl/hannahsten/texifyidea/settings/conventions/TexifyConventionsConfigurable.kt:
```diff
@@ -0,0 +1,216 @@
+
+        labelConventionsTable = createConventionsTable()
+
+        maxSectionSize = JBIntSpinner(4000, 100, Integer.MAX_VALUE)
```
This spinner isn't yet working? It resets if I try to change it

PHPirates replied:
Ah, no, it was resetting because I was providing a value lower than the minimum of 100. But it only resets when you click ok so that's very confusing, maybe just leave the minimum at 1

PHPirates reviewed (changes requested):
Looks like a good step in the right direction! I think the first thing to do now is make sure all the intended functionality, serving as a showcase for the schemes, is actually working.

PHPirates commented:
You are right, I forgot we can also add custom components there.

PHPirates commented on src/nl/hannahsten/texifyidea/psi/LatexEnvironmentUtil.kt:
```diff
@@ -23,7 +25,16 @@ fun getLabel(element: LatexEnvironment): String? {
-        if (!EnvironmentMagic.labeled.containsKey(element.environmentName)) return null
+        // Not very clean. We don't really need the conventions here, but determine which environments *can* have a
+        // label. However, if we didn't use the conventions, we would have to duplicate the information in
+        // EnvironmentMagic
```
The information wouldn't be duplicated, if everybody uses it from Magic.
What I did in a similar situation, is I kept the 'hardcoded' magic list, like the one that is now in TexifyConventionsScheme, in the Magic file. This ensures that everything magic is kept in Magic instead of spread around in various managers, while it also allows any code to access the hardcoded list instead of the user-changed list, which is useful if getting the second one has a significant performance impact.
I don't think there is any performance impact in this case, so I think this code is fine as it is.

PHPirates reviewed (approved):
Super! I think this is ready to merge, changing any of the conventions works for me.
This is a great contribution, it will make it so much easier to make magic user-configurable.
I will include it in an alpha release right away to try it out myself.

PHPirates commented:
Do you want to update the wiki yourself?

fberlakovich commented:
> (I just did the too large section inspection, indeed it's very easy to use)

Sorry, I completely missed it. Did the old setting for it even work because apparently the inspection didn't make use of it?

fberlakovich commented:
> Do you want to update the wiki yourself?

Sure, I will update it.

fberlakovich commented:
> Do you want to update the wiki yourself?

https://github.com/Hannah-Sten/TeXiFy-IDEA/wiki/Features#conventions

PHPirates commented:
> > (I just did the too large section inspection, indeed it's very easy to use)
> 
> Sorry, I completely missed it. Did the old setting for it even work because apparently the inspection didn't make use of it?

Don't think there was any setting.

> > Do you want to update the wiki yourself?
> 
> https://github.com/Hannah-Sten/TeXiFy-IDEA/wiki/Features#conventions

Great, thanks a lot! I also removed the old global sectioning level setting from the wiki.
